Allow request strategy to be passed to client on construction.

peterschutt opened the pull request:
For #58

Preserves the default behavior of `RequestStrategy` instance, however allows passing an alternate request strategy on construction of the client.

MikeWooster commented on apiclient/client.py:
```diff
@@ -21,6 +21,7 @@ def __init__(
         response_handler: Type[BaseResponseHandler] = RequestsResponseHandler,
         request_formatter: Type[BaseRequestFormatter] = NoOpRequestFormatter,
         error_handler: Type[BaseErrorHandler] = ErrorHandler,
+        request_strategy: Type[BaseRequestStrategy] = RequestStrategy,
```
```suggestion
        request_strategy: BaseRequestStrategy = RequestStrategy(),
```

peterschutt replied:
That's not compatible behavior with the current impl though because of the "function defaults evaluated at compile time" thing and that the strategy stores a copy of the client on itself. E.g.,:

This is current behavior:

```python
import apiclient
c1 = apiclient.APIClient()
c2 = apiclient.APIClient()
c1._request_strategy is c2._request_strategy  # False
c1._request_strategy._client is c2._request_strategy._client  # False
```

But if the default is instantiated in the function signature:

```python
import apiclient
c1 = apiclient.APIClient()
c2 = apiclient.APIClient()
c1._request_strategy is c2._request_strategy  # True
c1._request_strategy._client is c2._request_strategy._client  # True
```

Would it be better if we did this?

```python
   def __init__(
        self,
        ...,
        request_strategy: Optional[BaseRequestStrategy] = None,
    ):
        ...
        if request_strategy is None:
            request_strategy = RequestStrategy()
        self.set_request_strategy(request_strategy)
```

MikeWooster replied:
Okay, that's a fair point.  Your proposed `Optional` does work better.

MikeWooster commented on apiclient/client.py:
```diff
@@ -35,7 +36,7 @@ def __init__(
         self.set_error_handler(error_handler)
         self.set_request_formatter(request_formatter)
-        self.set_request_strategy(RequestStrategy())
+        self.set_request_strategy(request_strategy())
```
```suggestion
        self.set_request_strategy(request_strategy)
```

MikeWooster reviewed (commented):
One small change.  As the client needs an instance of the strategy, it would be better to pass in the initialized strategy.  This means that if for any reason you have extended it and added your own `__init__`, you can still pass it in to the client.

peterschutt commented:
Thanks for agreeing to add this.

As it is, the PR is quite a simple change - would you like to see any tests added? 

Also, while there is no problem with instantiating the `NoAuthentication` instance in the method signature, are you happy with the different patterns?

MikeWooster commented:
> Thanks for agreeing to add this.
> 
> As it is, the PR is quite a simple change - would you like to see any tests added?
> 
> Also, while there is no problem with instantiating the `NoAuthentication` instance in the method signature, are you happy with the different patterns?

I think it would be best if we use the `Optional` signature there too now we have added it here.

Maybe a single test should be added, which just tests that the `get_request_strategy` method on the client returns the strategy you passed in.

peterschutt commented:
- changed client signature to `authentication_method: Optional[BaseAuthenticationMethod] = None`
- modified `test_client_initialization_with_invalid_authentication_method()` as `None` has become a valid value for `authentication_method`.
- added `test_client_initialization_with_invalid_request_strategy()` which tests the raised runtime error like the tests for the other args.
- added `test_authentication_method_with_user_defined()` and `test_get_request_strategy_with_user_defined()` which test the relevant getter methods return the instance passed on instantiation.
